### DISCUSSION TITLE
Question: Draggable Pie Chart

### DISCUSSION THREAD

**Discussion by ChristianRaoulis** [antvis/G2#5473] [Q&A] [answered]
I would love to create a Draggable Pie Chart ([example](https://www.jamesalvarez.co.uk/blog/draggable-piechart-js-class/)) with G2 but I got no idea if something like that is even possible with G2.

Is that even possible?
If yes, how would i do something like that?

**Comment by hustcc** [comment]
I'm trying to create a pie chart like this, but the challenge is calculating the new position of the sector when dragging the interactive point based on the angle.

Since it's difficult to determine the current angle of the sector and the angle after dragging the offset x y, it's best to create a custom shape for this pie chart. In this custom shape, you can draw the sector and the draggable point, and have access to more contextual information about the graphics.

**Comment by ChristianRaoulis** [comment]
is there an example for that somewhere? Cuz i got no clue where to start 😅

**Comment by pearmini** [comment]
I'll try to write a demo this week.

**Comment by pearmini** [comment, marked as answer]
@Skyriis  Here is a [demo](https://observablehq.com/@pearmini/draggable-pie-chart-for-g2) for it.

**Reply by hustcc** [reply]
Can this interaction be written as a custom interaction? and register it into G2.

```ts
import { register } from '@antv/g2'; 
register('interaction.dragpie', () => {});

chart.interaction('dragpie', {});
```

**Reply by pearmini** [reply]
Yes, but I'm not sure the interface of interaction component is stable. It maybe change in the future?

**Reply by ChristianRaoulis** [reply]
how would i add labels to the chart parts? :o

**Reply by pearmini** [reply]
You can see this example：https://g2.antv.antgroup.com/examples/general/pie/#pie

**Comment by ChristianRaoulis** [comment]
i'm running into an issue with the autofit setting and i got no clue what causes this :o

https://github.com/antvis/G2/assets/128689017/7ab75f38-a299-44e9-9221-7daed11624f0

Any ideas how to fix that?
Why does it even move down and why does it get bigger?

**Reply by pearmini** [reply]
You should clear the old interaction and apply new interaction when chart resizing:

```js
function useDragPie(chart) {}

const [apply, clear] = useDragPie(chart);

chart.on(ChartEvent.afterrender, () => {
   clear();
   apply();
});
```

**Reply by ChristianRaoulis** [reply]
could you show that code in the demo as an example? 

I still don't fully got what i need to do 😅

**Reply by ChristianRaoulis** [reply]
i fixed it with a `window.addEventListener('resize')` which removes the control points. Thank you

**Comment by ChristianRaoulis** [comment]
I got another two issues with the chart.

1. a pie element somehow can skip its neighbour 

https://github.com/antvis/G2/assets/128689017/78a98ea8-c667-4a7f-9b09-d7d984b3128d

2. a pie element can not be bigger then 50% of the chart

https://github.com/antvis/G2/assets/128689017/9f66c1fd-a9e3-48fa-aa62-0022befa1e2c


Is there a way to fix that?

**Reply by pearmini** [reply]
- The first is a bug which related the code below, but I do not have enough time to fix them. 

![image](https://github.com/antvis/G2/assets/49330279/29b7e1b9-1803-454d-ba75-f12d6113cc1f)

- The second is a feature missing which you can try you self.

**Reply by ChristianRaoulis** [reply]
> * The first is a bug which related the code below, but I do not have enough time to fix them.

Thanks for the hint. I fixed it by replacing the if with `if (c1 * c2 < 0 || (c1 <= 0 && c2 <= 0)) return;`

**Reply by ChristianRaoulis** [reply]
> * The second is a feature missing which you can try you self.

tbh i got no idea where to start
